contracts/BookMarketplace.sol

```diff
@@ -44,15 +48,98 @@ contract BookMarketplace is ERC721URIStorage {
         bool sold
     );
 
+    event ListingUpdated(
+        uint256 indexed tokenId,
+        address seller,
+        address owner,
+        uint256 instantPrice,
+        uint256 startingPrice,
+        bool allowBid,
+        Bid[] bidList,
+        bool sold
+    );
+
+    /* modifiers */
+    
+    
+    // Modifier prevents function from being called by anyone other than the market owner
+    modifier onlyMarketOwner() {
+        require(
+            marketOwner == msg.sender,
+            "Only the marketplace owner can call this function."
+        );
+        _;
+    }
+
+    /* constructor */
+
     constructor() ERC721("BookExchange Tokens", "BOOK") {
-        owner = payable(msg.sender);
+        marketOwner = payable(msg.sender);
+    }
+
+    /* functions */
+
+    function getMarketOwnerAddress() public view returns(address) {
+        return marketOwner;
+    }
+
+    // change the price of a listing
+    function updateInstantPrice(uint256 tokenId, uint256 newPrice) public payable onlyMarketOwner {
+        Listing storage listing = idToMarketItem[tokenId];
+        listing.instantPrice = newPrice;
+        idToMarketItem[tokenId] = listing;
+    }
+
+    // get the price of a listing
+    function getInstantPrice(uint256 tokenId) public view returns (uint256) {
+        Listing storage listing = idToMarketItem[tokenId];
+        return listing.instantPrice;
+    }
+
+    // Adds bids to a bid list by tokenId
+    function addBid(uint256 tokenId, uint256 bidAmount) public payable {
+        Listing storage listing = idToMarketItem[tokenId];
+        require(
+            listing.allowBid == true,
+            "This item does not allow bids."
+        );
+        require(
+            listing.sold == false,
+            "This item has been sold."
+        );
+        require(
+            bidAmount >= listing.instantPrice,
+            "Bid must be greater than or equal to the current price."
+        );
+        require(
+            msg.value == bidAmount,
+            "Bid amount must be equal to the amount sent." 
+        );
+        Bid memory newBid;
+        newBid.bidder = msg.sender;
+        newBid.bidAmount = bidAmount;
+        listing.bidList.push(newBid);
+        idToMarketItem[tokenId] = listing;
```

Comment:
No need to put listing back to idToMarketItem. When you push the newBid, it's already modified.

Reply:
Hm, I see, Ill change this and retest.